compaction/compaction.cc

```diff
@@ -220,13 +220,13 @@ class compaction_write_monitor final : public sstables::write_monitor, public ba
 
     ~compaction_write_monitor() {
         if (_sst) {
-            _table_s.get_compaction_strategy().get_backlog_tracker().revert_charges(_sst);
+            _table_s.get_backlog_tracker().revert_charges(_sst);
```

Comment:
Q: wouldn't we have one compaction strategy per compaction_group? Then the original code still works.

Reply:
A: no, compaction_strategy is stateless wrt the table itself (it only holds compaction strategy parameters).

Reply:
Ah you have a good point there. each group should have its own compaction strategy instance, because the strategy holds state which shouldn't be shared across groups. Anyway, I think we're moving towards a good direction with compaction manager owning the backlog tracker, allowing the manager for example to inspect the backlog of each table (group) in the future for improving decision making, and also complete the confinement of the concept to the compaction namespace.

Reply:
TWCS state, for example:

```
class time_window_compaction_strategy : public compaction_strategy_impl {
    time_window_compaction_strategy_options _options;
    int64_t _estimated_remaining_tasks = 0;
    db_clock::time_point _last_expired_check;
    // As timestamp_type is an int64_t, a primitive type, it must be initialized here.
    timestamp_type _highest_window_seen = 0;
    // Keep track of all recent active windows that still need to be compacted into a single SSTable
    std::unordered_set<timestamp_type> _recent_active_windows;
    size_tiered_compaction_strategy_options _stcs_options;
```

_last_expired_check, _highest_window_seen, _recent_active_windows are states that depend on SSTable set, therefore belongs to a single group.

I'll send another series to have one compaction_strategy per group, following this one. compaction_strategy is being exposed through table_state, so we can easily have one per group. what do u think?

Reply:
We can consider moving in two directions:
1. proclaim that it's bad that the strategy holds per-table state, and instead add a method compaction_strategy::create_state() (and ask everyone to pass the state on every call). Then compaction_strategy can be shared across groups. This fits will with the name (a strategy is an algorithm that applies to state, not algorithm+state).
2. embrace the current situation and have one strategy per group.

Reply:
I think the first one is cleaner, but you're in a better position to estimate the churn involved.

Reply:
> I think the first one is cleaner, but you're in a better position to estimate the churn involved.

I think the first one fits better with the whole design. compaction_strategy is a table property that must be shared among all groups. and only the state of a group should be made local to it.